src/libcore/num/cast.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use fmt;
+
+/// Internal trait for APIs like `i32::cast`.
+#[unstable(feature = "num_cast_internals", issue = "0")]
+pub trait Cast<T>: Sized {
+    /// Internal implementation detail of this trait.
+    fn cast(t: T) -> Result<Self, CastError>;
+}
+
+/// Error type returned from APIs like `i32::cast`, indicates that a cast could
+/// not be performed losslessly.
+#[unstable(feature = "num_cast", issue = "0")]
+#[derive(Debug)]
+pub struct CastError(());
+
+#[unstable(feature = "num_cast", issue = "0")]
+impl fmt::Display for CastError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        "failed to losslessly cast integral types".fmt(f)
+    }
+}
+
+macro_rules! same_sign_cast_int_impl {
+    ($storage:ty, $target:ty, $($source:ty),*) => {$(
+        #[unstable(feature = "num_cast", issue = "0")]
+        impl Cast<$source> for $target {
+            fn cast(u: $source) -> Result<$target, CastError> {
```

Comment:
These functions want an `#[inline]`. I’ve observed in my Range specialisation PR that the casts won’t get inlined and inlining is able to get rid of the unnecessary code resulting in one or two instructions per cast.